rfcs/20221006-frontend_event_tracking.md

```diff
@@ -0,0 +1,469 @@
+
+## Reviewers
+
+- [ ] @krysal - For API experience
+- [x] @dhruvkb - For frontend and DX experience
+
+## Rationale
+
+Openverse contributors have very little insight into the behaviors and interests of our users. We do not have any means of observing if they actually use the features we build in the intended manner, or at all.
+
+There are multiple ways to address this, like conducting user surveys or test sessions. One other way we would like to gather this information is through analytics, specifically logging events performed by users. I'll share a full list of the events later in this proposal, but here are some examples:
+
+- When a user clicks on a search result from the search results page
+- When a user copies the attribution text
+
+By collecting this data, anonymized and with user consent (through opt-out), we can make better decisions about what to build and increase the overall usefulness of Openverse.
+
+### Why is this distinct from a general-purpose, cross-repo Analytics RFC?
+
+This proposal only discusses front-end event collection. It does not discuss:
+
+- Dashboards or tools for _analysing_ collected data
+- The ideal data warehouse for our frontend events
+- Whether to use a SaaS product or self-host analytics software
+
+I suggest that our frontend events should be agnostic to any specific analytics platform. This means that our frontend events will be generalized to contain information we think is useful, and munged as necessary to conform to the needs of our analytics service. I see many benefits to doing this:
+
+- It avoids vendor lock-in to a specific analytics platform
+- It makes it much easier for users to audit exactly what values we're collecting, by avoiding heavy and mysterious 3rd-party SDKs
+- Makes us less likely to be blocked by ad and script blocking software
+
+There are many established conventions around frontend event collection. Using our own system should allow us to adopt most 3rd party platforms without locking us in. A quick example use case is the Google Analytics alternative [Plausible.io](https://plausible.io/), which includes an [events API](https://plausible.io/docs/events-api). Plausible accepts the following simple request shape to capture a pageview, for example:
+
+```shell
+curl -i -X POST https://plausible.io/api/event \
+  -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36 OPR/71.0.3770.284' \
+  -H 'X-Forwarded-For: 127.0.0.1' \
+  -H 'Content-Type: application/json' \
+  --data '{"name":"pageview","url":"http://dummy.site","domain":"dummy.site","screen_width":1666}'
+```
+
+There are many points in the code execution at which we could transform our custom events into ones supported by other platforms. By adopting and implementing our own custom solution, we can start collecting data as soon as possible without being blocked on time spent evaluating and implementing a full analytics system.
```

Comment:
Is there a way to surface the implicit infrastructure decisions here? For example, where will the event data be stored and what limitations might that place on future plans?

Reply:
I've commented [here](https://github.com/WordPress/openverse/pull/312#discussion_r1062735808) with some context about this question, thanks for asking!